File: ios/Plugin/StrataStoragePlugin.swift
import Foundation
import Capacitor

/**
 * Main Capacitor plugin for Strata Storage
 * Coordinates between different storage types on iOS.
 *
 * NOTE: Method registration with the Capacitor bridge lives in
 * StrataStoragePlugin.m (CAP_PLUGIN / CAP_PLUGIN_METHOD). Keep the two in
 * sync when adding or renaming methods.
 */
@objc(StrataStoragePlugin)
public class StrataStoragePlugin: CAPPlugin {
    private let userDefaultsStorage = UserDefaultsStorage()
    private let keychainStorage = KeychainStorage()
    private let filesystemStorage = FilesystemStorage()

    /// Storage types that have a real native backend on iOS.
    private let supportedStorageTypes: Set<String> = ["preferences", "secure", "sqlite", "filesystem"]

    // MARK: - SQLite multi-store helpers

    /// Default database name (→ file `strata_storage.db`) and table.
    private static let defaultDatabase = "strata_storage"
    private static let defaultTable = "storage"

    /// Allowed characters for a SQLite identifier body (the first character must
    /// additionally not be a digit).
    private static let sqliteIdentifierChars =
        CharacterSet(charactersIn: "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_")

    /// Validates a `database`/`table` identifier against `^[A-Za-z_][A-Za-z0-9_]*$`.
    /// These two cannot be bound as parameters (they are interpolated into SQL /
    /// the `.db` filename), so they are the single injection surface; every
    /// other value is bound. Non-conforming identifiers are REJECTED rather than
    /// silently stripped — the TS adapter applies the same allow-list and rejects
    /// first, so this is defense-in-depth (and also blocks path traversal).
    private func isValidSqliteIdentifier(_ s: String) -> Bool {
        guard let first = s.unicodeScalars.first else { return false }
        if CharacterSet(charactersIn: "0123456789").contains(first) { return false }
        return s.unicodeScalars.allSatisfy { StrataStoragePlugin.sqliteIdentifierChars.contains($0) }
    }

    /// Returns the validated table identifier for the call, or nil after
    /// rejecting `call` with a clear error.
    private func validatedTable(_ call: CAPPluginCall) -> String? {
        let table = call.getString("table") ?? StrataStoragePlugin.defaultTable
        guard isValidSqliteIdentifier(table) else {
            call.reject("Invalid SQLite table name: \(table)")
            return nil
        }
        return table
    }

    /// Returns the cached SQLite store for the call's validated `database`, or
    /// nil after rejecting `call`. The filename cannot be a bound parameter, so
    /// the name is validated before `.db` is appended.
    private func validatedSqliteStore(for call: CAPPluginCall) -> SQLiteStorage? {
        let database = call.getString("database") ?? StrataStoragePlugin.defaultDatabase
        guard isValidSqliteIdentifier(database) else {
            call.reject("Invalid SQLite database name: \(database)")
            return nil
        }
        return SQLiteStorageManager.shared.store(forFile: "\(database).db")
    }

    /**
     * Check if a specific storage type is available.
     * Matches the JS contract: resolves `{ available: boolean }`.
     */
    @objc func isAvailable(_ call: CAPPluginCall) {
        let storage = call.getString("storage") ?? "preferences"
        let available: Bool
        switch storage {
        case "sqlite":
            let database = call.getString("database") ?? StrataStoragePlugin.defaultDatabase
            available = isValidSqliteIdentifier(database)
                && SQLiteStorageManager.shared.store(forFile: "\(database).db").isOpen
        case "filesystem":
            available = filesystemStorage.isAvailable()
        case "preferences", "secure":
            available = true
        default:
            available = false
        }
        call.resolve([
            "available": available
        ])
    }

    /**
     * Get value from storage.
     * For sqlite/filesystem the resolved `value` is the full StorageValue
     * wrapper object (or NSNull on a miss).
     */
    @objc func get(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else {
            call.reject("Key is required")
            return
        }

        let storage = call.getString("storage") ?? "preferences"

        do {
            let value: Any?

            switch storage {
            case "secure":
                value = try keychainStorage.get(key: key)
            case "sqlite":
                guard let store = validatedSqliteStore(for: call), let table = validatedTable(call) else { return }
                value = store.get(table: table, key: key)
            case "filesystem":
                value = filesystemStorage.get(key: key)
            case "preferences":
                fallthrough
            default:
                value = userDefaultsStorage.get(key: key)
            }

            call.resolve([
                "value": value ?? NSNull()
            ])
        } catch {
            call.reject("Failed to get value", nil, error)
        }
    }

    /**
     * Set value in storage.
     * For sqlite/filesystem the `value` option is the full StorageValue
     * wrapper object and is stored verbatim (JSON) for a perfect round-trip.
     */
    @objc func set(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else {
            call.reject("Key is required")
            return
        }

        let storage = call.getString("storage") ?? "preferences"

        do {
            switch storage {
            case "secure":
                let value = call.getValue("value") ?? NSNull()
                _ = try keychainStorage.set(key: key, value: value)
            case "sqlite":
                guard let store = validatedSqliteStore(for: call), let table = validatedTable(call) else { return }
                // getObject returns JSObject ([String: JSValue]); upcast each
                // value to Any (the canonical Capacitor pattern) so the bridged
                // Foundation values are JSONSerialization-compatible.
                guard let jsObject = call.getObject("value") else {
                    call.reject("SQLite value must be a StorageValue object")
                    return
                }
                let ok = try store.set(
                    table: table,
                    key: key,
                    wrapper: jsObject as [String: Any]
                )
                if !ok {
                    call.reject("Failed to write value to SQLite")
                    return
                }
            case "filesystem":
                guard let jsObject = call.getObject("value") else {
                    call.reject("Filesystem value must be a StorageValue object")
                    return
                }
                let ok = try filesystemStorage.set(key: key, wrapper: jsObject as [String: Any])
                if !ok {
                    call.reject("Failed to write value to filesystem")
                    return
                }
            case "preferences":
                fallthrough
            default:
                let value = call.getValue("value") ?? NSNull()
                userDefaultsStorage.set(key: key, value: value)
            }

            call.resolve()
        } catch {
            call.reject("Failed to set value", nil, error)
        }
    }

    /**
     * Remove value from storage
     */
    @objc func remove(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else {
            call.reject("Key is required")
            return
        }

        let storage = call.getString("storage") ?? "preferences"

        do {
            switch storage {
            case "secure":
                _ = try keychainStorage.remove(key: key)
            case "sqlite":
                guard let store = validatedSqliteStore(for: call), let table = validatedTable(call) else { return }
                _ = store.remove(table: table, key: key)
            case "filesystem":
                _ = filesystemStorage.remove(key: key)
            case "preferences":
                fallthrough
            default:
                userDefaultsStorage.remove(key: key)
            }

            call.resolve()
        } catch {
            call.reject("Failed to remove value", nil, error)
        }
    }

    /**
     * Clear storage.
     * The JS contract sends an optional `pattern`. The native backends accept
     * a prefix; we use `pattern` as that prefix (prefix/contains matching).
     */
    @objc func clear(_ call: CAPPluginCall) {
        let storage = call.getString("storage") ?? "preferences"
        // Accept both `pattern` (JS contract) and legacy `prefix`.
        let prefix = call.getString("pattern") ?? call.getString("prefix")

        do {
            switch storage {
            case "secure":
                _ = try keychainStorage.clear(prefix: prefix)
            case "sqlite":
                guard let store = validatedSqliteStore(for: call), let table = validatedTable(call) else { return }
                _ = store.clear(table: table, prefix: prefix)
            case "filesystem":
                _ = filesystemStorage.clear(prefix: prefix)
            case "preferences":
                fallthrough
            default:
                userDefaultsStorage.clear(prefix: prefix)
            }

            call.resolve()
        } catch {
            call.reject("Failed to clear storage", nil, error)
        }
    }

    /**
     * Get all keys from storage
     */
    @objc func keys(_ call: CAPPluginCall) {
        let storage = call.getString("storage") ?? "preferences"
        let pattern = call.getString("pattern")

        do {
            let keys: [String]

            switch storage {
            case "secure":
                keys = try keychainStorage.keys(pattern: pattern)
            case "sqlite":
                guard let store = validatedSqliteStore(for: call), let table = validatedTable(call) else { return }
                keys = store.keys(table: table, pattern: pattern)
            case "filesystem":
                keys = filesystemStorage.keys(pattern: pattern)
            case "preferences":
                fallthrough
            default:
                keys = userDefaultsStorage.keys(pattern: pattern)
            }

            call.resolve([
                "keys": keys
            ])
        } catch {
            call.reject("Failed to get keys", nil, error)
        }
    }

    /**
     * Get storage size information.
     * When `detailed` is true, resolves `{ total, count, detailed: { keys,
     * values, metadata } }`; otherwise `{ total, count }`.
     */
    @objc func size(_ call: CAPPluginCall) {
        let storage = call.getString("storage") ?? "preferences"
        let detailed = call.getBool("detailed") ?? false

        do {
            // segments carries total/count/keys/values/metadata in bytes.
            let segments: [String: Int]

            switch storage {
            case "secure":
                let info = try keychainStorage.size()
                segments = ["total": info.total, "count": info.count, "keys": 0, "values": info.total, "metadata": 0]
            case "sqlite":
                guard let store = validatedSqliteStore(for: call), let table = validatedTable(call) else { return }
                segments = try store.size(table: table)
            case "filesystem":
                segments = filesystemStorage.size()
            case "preferences":
                fallthrough
            default:
                let info = userDefaultsStorage.size()
                segments = ["total": info.total, "count": info.count, "keys": 0, "values": info.total, "metadata": 0]
            }

            var result: [String: Any] = [
                "total": segments["total"] ?? 0,
                "count": segments["count"] ?? 0
            ]
            if detailed {
                result["detailed"] = [
                    "keys": segments["keys"] ?? 0,
                    "values": segments["values"] ?? 0,
                    "metadata": segments["metadata"] ?? 0
                ]
            }
            call.resolve(result)
        } catch {
            call.reject("Failed to get size", nil, error)
        }
    }

    /**
     * Query SQLite-backed storage.
     * Matches the optional `query` method in the JS contract: resolves
     * `{ results: [{ key, value }] }` where `value` is the full StorageValue
     * wrapper for each non-expired row, so the JS adapter applies the condition
     * in one round-trip without re-fetching each key.
     */
    @objc func query(_ call: CAPPluginCall) {
        let storage = call.getString("storage") ?? "sqlite"

        guard storage == "sqlite" else {
            call.reject("Query is only supported for the 'sqlite' storage type")
            return
        }

        guard let store = validatedSqliteStore(for: call), let table = validatedTable(call) else { return }
        let condition = call.getObject("condition") ?? [:]
        let results = store.query(table: table, condition: condition)
        call.resolve([
            "results": results
        ])
    }

    /**
     * Delete every expired row for a SQLite store and resolve `{ removed }`.
     * The JS adapter calls this on its TTL cleanup tick, since `keys`/`query`
     * now exclude expired rows in SQL (no lazy per-key deletion on read).
     */
    @objc func cleanupExpired(_ call: CAPPluginCall) {
        let storage = call.getString("storage") ?? "sqlite"

        guard storage == "sqlite" else {
            call.reject("cleanupExpired is only supported for the 'sqlite' storage type")
            return
        }

        guard let store = validatedSqliteStore(for: call), let table = validatedTable(call) else { return }
        let removed = store.cleanupExpired(table: table)
        call.resolve([
            "removed": removed
        ])
    }

    /**
     * iOS-specific: read a UserDefaults value.
     * Resolves `{ value: unknown }`.
     */
    @objc func getUserDefaults(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else {
            call.reject("Key is required")
            return
        }
        let suiteName = call.getString("suiteName")
        let store = suiteName != nil ? UserDefaultsStorage(suiteName: suiteName) : userDefaultsStorage
        call.resolve([
            "value": store.get(key: key) ?? NSNull()
        ])
    }

    /**
     * iOS-specific: write a UserDefaults value.
     */
    @objc func setUserDefaults(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else {
            call.reject("Key is required")
            return
        }
        let value = call.getValue("value") ?? NSNull()
        let suiteName = call.getString("suiteName")
        let store = suiteName != nil ? UserDefaultsStorage(suiteName: suiteName) : userDefaultsStorage
        _ = store.set(key: key, value: value)
        call.resolve()
    }

    /**
     * iOS-specific: read a Keychain value. Resolves `{ value: string | null }`.
     */
    @objc func getKeychain(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else {
            call.reject("Key is required")
            return
        }
        let service = call.getString("service")
        let accessGroup = call.getString("accessGroup")
        let store = (service != nil || accessGroup != nil)
            ? KeychainStorage(service: service, accessGroup: accessGroup)
            : keychainStorage

        do {
            let value = try store.get(key: key)
            // The JS contract types this as `string | null`. Serialize
            // non-string values to a JSON string so the bridge stays typed.
            if let value = value {
                if let str = value as? String {
                    call.resolve(["value": str])
                } else if let data = try? JSONSerialization.data(withJSONObject: value, options: []),
                          let str = String(data: data, encoding: .utf8) {
                    call.resolve(["value": str])
                } else {
                    call.resolve(["value": NSNull()])
                }
            } else {
                call.resolve(["value": NSNull()])
            }
        } catch {
            call.reject("Failed to read keychain item", nil, error)
        }
    }

    /**
     * iOS-specific: write a Keychain value, honoring the `accessible` option.
     */
    @objc func setKeychain(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else {
            call.reject("Key is required")
            return
        }
        guard let value = call.getString("value") else {
            call.reject("Value is required and must be a string")
            return
        }
        let service = call.getString("service")
        let accessGroup = call.getString("accessGroup")
        let accessible = call.getString("accessible")
        let store = (service != nil || accessGroup != nil)
            ? KeychainStorage(service: service, accessGroup: accessGroup)
            : keychainStorage

        do {
            _ = try store.set(key: key, value: value, accessible: accessible)
            call.resolve()
        } catch {
            call.reject("Failed to write keychain item", nil, error)
        }
    }
}

File: android/src/main/java/com/stratastorage/StrataStoragePlugin.java
package com.stratastorage;

import com.getcapacitor.JSObject;
import com.getcapacitor.JSArray;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.strata.storage.SharedPreferencesStorage;
import com.strata.storage.EncryptedStorage;
import com.strata.storage.FilesystemStorage;
import com.strata.storage.SQLiteStorage;
import android.util.Log;
import org.json.JSONArray;
import org.json.JSONException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

/**
 * Main Capacitor plugin for Strata Storage
 * Coordinates between different storage types on Android.
 *
 * Storage types with a real native backend: preferences, secure, sqlite,
 * filesystem.
 *
 * SQLite supports multi-store: the `database` option selects the DB file and
 * `table` selects the table within it. SQLite helper instances are cached per
 * database name so the connection is not reopened per call.
 */
@CapacitorPlugin(name = "StrataStorage")
public class StrataStoragePlugin extends Plugin {
    private static final String DEFAULT_DATABASE = "strata_storage";
    private static final String DEFAULT_TABLE = "storage";

    /** Strict SQL-identifier allow-list for the database file stem. */
    private static final Pattern SQLITE_IDENTIFIER = Pattern.compile("^[A-Za-z_][A-Za-z0-9_]*$");

    private SharedPreferencesStorage sharedPrefsStorage;
    private EncryptedStorage encryptedStorage;
    private FilesystemStorage filesystemStorage;

    /** SQLite helper cache keyed by resolved database name (without ".db"). */
    private final Map<String, SQLiteStorage> sqliteStores = new HashMap<>();

    @Override
    public void load() {
        try {
            sharedPrefsStorage = new SharedPreferencesStorage(getContext());
        } catch (Exception e) {
            Log.e("StrataStorage", "Failed to initialize preferences storage", e);
        }
        try {
            encryptedStorage = new EncryptedStorage(getContext());
        } catch (Exception e) {
            // Secure storage may be unavailable on some devices/keystores —
            // keep the rest of the plugin functional.
            Log.e("StrataStorage", "Failed to initialize encrypted storage", e);
        }
        try {
            filesystemStorage = new FilesystemStorage(getContext());
        } catch (Exception e) {
            Log.e("StrataStorage", "Failed to initialize filesystem storage", e);
        }
    }

    /**
     * Validate a database name as a strict SQL identifier
     * ({@code ^[A-Za-z_][A-Za-z0-9_]*$}) and return it as the file stem. The
     * ".db" suffix is appended when opening. A null/empty name resolves to the
     * default; any other non-conforming name is REJECTED rather than silently
     * stripped (the TS adapter applies the same allow-list and rejects first —
     * this is defense-in-depth, and also blocks path separators / traversal).
     */
    private static String sanitizeDatabase(String database) {
        if (database == null || database.isEmpty()) {
            return DEFAULT_DATABASE;
        }
        if (!SQLITE_IDENTIFIER.matcher(database).matches()) {
            throw new IllegalArgumentException("Invalid SQLite database name: " + database);
        }
        return database;
    }

    /**
     * Return (creating + caching on first use) the SQLite helper for the given
     * database name. The cache is keyed by the validated DB stem so the
     * connection is reused across bridge calls. Returns {@code null} if the name
     * is invalid or the helper cannot be constructed — callers reject the call.
     */
    private synchronized SQLiteStorage getSqliteStore(String database) {
        final String dbStem;
        try {
            dbStem = sanitizeDatabase(database);
        } catch (IllegalArgumentException e) {
            Log.e("StrataStorage", "Invalid SQLite database name", e);
            return null;
        }
        SQLiteStorage store = sqliteStores.get(dbStem);
        if (store == null) {
            try {
                store = new SQLiteStorage(getContext(), dbStem + ".db");
                sqliteStores.put(dbStem, store);
            } catch (Exception e) {
                Log.e("StrataStorage", "Failed to open SQLite database " + dbStem, e);
                return null;
            }
        }
        return store;
    }

    /**
     * Check if a specific storage type is available.
     * Matches the JS contract: resolves { available: boolean }.
     */
    @PluginMethod
    public void isAvailable(PluginCall call) {
        String storage = call.getString("storage", "preferences");
        boolean available;
        switch (storage) {
            case "secure":
                available = encryptedStorage != null;
                break;
            case "sqlite":
                available = getSqliteStore(call.getString("database", DEFAULT_DATABASE)) != null;
                break;
            case "filesystem":
                available = filesystemStorage != null;
                break;
            case "preferences":
            default:
                available = sharedPrefsStorage != null;
                break;
        }

        JSObject result = new JSObject();
        result.put("available", available);
        call.resolve(result);
    }

    /**
     * Get value from storage.
     *
     * For sqlite/filesystem the resolved `value` is the full wrapper object
     * (parsed back from JSON). A miss resolves `value` = null.
     */
    @PluginMethod
    public void get(PluginCall call) {
        String key = call.getString("key");
        if (key == null) {
            call.reject("Key is required");
            return;
        }

        String storage = call.getString("storage", "preferences");

        try {
            switch (storage) {
                case "secure": {
                    if (encryptedStorage == null) {
                        call.reject("Encrypted storage not available");
                        return;
                    }
                    resolveValue(call, encryptedStorage.get(key));
                    return;
                }
                case "sqlite": {
                    SQLiteStorage store = getSqliteStore(call.getString("database", DEFAULT_DATABASE));
                    if (store == null) {
                        call.reject("SQLite storage not available");
                        return;
                    }
                    String table = call.getString("table", DEFAULT_TABLE);
                    resolveValue(call, store.get(table, key));
                    return;
                }
                case "filesystem": {
                    if (filesystemStorage == null) {
                        call.reject("Filesystem storage not available");
                        return;
                    }
                    resolveValue(call, filesystemStorage.get(key));
                    return;
                }
                case "preferences":
                default: {
                    if (sharedPrefsStorage == null) {
                        call.reject("Preferences storage not available");
                        return;
                    }
                    resolveValue(call, sharedPrefsStorage.get(key));
                    return;
                }
            }
        } catch (Exception e) {
            call.reject("Failed to get value", e);
        }
    }

    /**
     * Resolve a get() call, mapping a Java {@code null} to a JS {@code null}
     * value so the TS adapters treat it as a miss.
     */
    private void resolveValue(PluginCall call, Object value) {
        JSObject result = new JSObject();
        if (value == null) {
            result.put("value", JSObject.NULL);
        } else if (value instanceof String) {
            // preferences/secure persist the full StorageValue wrapper as a JSON
            // string; parse it back to an object so TTL/tags/metadata round-trip
            // (matching the sqlite/filesystem contract). Legacy/plain strings are
            // returned unchanged.
            result.put("value", parseWrapperOrRaw((String) value));
        } else {
            result.put("value", value);
        }
        call.resolve(result);
    }

    /** Parse a stored JSON-object wrapper string back into a {@link JSObject};
     * return the raw string for non-JSON-object payloads. */
    private Object parseWrapperOrRaw(String raw) {
        String trimmed = raw.trim();
        if (trimmed.startsWith("{")) {
            try {
                return new JSObject(raw);
            } catch (JSONException ignored) {
                // Not a JSON object — fall through and return the raw string.
            }
        }
        return raw;
    }

    /**
     * Set value in storage.
     *
     * For sqlite/filesystem the `value` is the FULL wrapper object
     * ({ value, created, updated, expires?, tags?, metadata? }); it is read via
     * call.getObject("value") and stored verbatim (with TTL/query columns
     * extracted for sqlite). For preferences/secure the raw value is forwarded
     * unchanged.
     */
    @PluginMethod
    public void set(PluginCall call) {
        String key = call.getString("key");
        if (key == null) {
            call.reject("Key is required");
            return;
        }

        String storage = call.getString("storage", "preferences");

        try {
            boolean ok;
            switch (storage) {
                case "secure":
                    if (encryptedStorage == null) {
                        call.reject("Encrypted storage not available");
                        return;
                    }
                    ok = encryptedStorage.set(key, call.getData().opt("value"));
                    break;
                case "sqlite": {
                    SQLiteStorage store = getSqliteStore(call.getString("database", DEFAULT_DATABASE));
                    if (store == null) {
                        call.reject("SQLite storage not available");
                        return;
                    }
                    JSObject wrapper = call.getObject("value");
                    if (wrapper == null) {
                        call.reject("A wrapper object 'value' is required for sqlite storage");
                        return;
                    }
                    ok = store.set(call.getString("table", DEFAULT_TABLE), key, wrapper);
                    break;
                }
                case "filesystem": {
                    if (filesystemStorage == null) {
                        call.reject("Filesystem storage not available");
                        return;
                    }
                    JSObject wrapper = call.getObject("value");
                    if (wrapper == null) {
                        call.reject("A wrapper object 'value' is required for filesystem storage");
                        return;
                    }
                    ok = filesystemStorage.set(key, wrapper);
                    break;
                }
                case "preferences":
                default:
                    if (sharedPrefsStorage == null) {
                        call.reject("Preferences storage not available");
                        return;
                    }
                    ok = sharedPrefsStorage.set(key, call.getData().opt("value"));
                    break;
            }

            if (!ok) {
                call.reject("Failed to persist value to " + storage + " storage");
                return;
            }
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to set value", e);
        }
    }

    /**
     * Remove value from storage
     */
    @PluginMethod
    public void remove(PluginCall call) {
        String key = call.getString("key");
        if (key == null) {
            call.reject("Key is required");
            return;
        }

        String storage = call.getString("storage", "preferences");

        try {
            switch (storage) {
                case "secure":
                    if (encryptedStorage == null) {
                        call.reject("Encrypted storage not available");
                        return;
                    }
                    encryptedStorage.remove(key);
                    break;
                case "sqlite": {
                    SQLiteStorage store = getSqliteStore(call.getString("database", DEFAULT_DATABASE));
                    if (store == null) {
                        call.reject("SQLite storage not available");
                        return;
                    }
                    store.remove(call.getString("table", DEFAULT_TABLE), key);
                    break;
                }
                case "filesystem":
                    if (filesystemStorage == null) {
                        call.reject("Filesystem storage not available");
                        return;
                    }
                    filesystemStorage.remove(key);
                    break;
                case "preferences":
                default:
                    if (sharedPrefsStorage == null) {
                        call.reject("Preferences storage not available");
                        return;
                    }
                    sharedPrefsStorage.remove(key);
                    break;
            }

            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to remove value", e);
        }
    }

    /**
     * Clear storage.
     * Accepts the JS `pattern` (used as a key prefix) and a legacy `prefix`.
     */
    @PluginMethod
    public void clear(PluginCall call) {
        String storage = call.getString("storage", "preferences");
        String prefix = call.getString("pattern");
        if (prefix == null) {
            prefix = call.getString("prefix");
        }

        try {
            switch (storage) {
                case "secure":
                    if (encryptedStorage == null) {
                        call.reject("Encrypted storage not available");
                        return;
                    }
                    encryptedStorage.clear(prefix);
                    break;
                case "sqlite": {
                    SQLiteStorage store = getSqliteStore(call.getString("database", DEFAULT_DATABASE));
                    if (store == null) {
                        call.reject("SQLite storage not available");
                        return;
                    }
                    store.clear(call.getString("table", DEFAULT_TABLE), prefix);
                    break;
                }
                case "filesystem":
                    if (filesystemStorage == null) {
                        call.reject("Filesystem storage not available");
                        return;
                    }
                    filesystemStorage.clear(prefix);
                    break;
                case "preferences":
                default:
                    if (sharedPrefsStorage == null) {
                        call.reject("Preferences storage not available");
                        return;
                    }
                    sharedPrefsStorage.clear(prefix);
                    break;
            }

            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to clear storage", e);
        }
    }

    /**
     * Get all keys from storage
     */
    @PluginMethod
    public void keys(PluginCall call) {
        String storage = call.getString("storage", "preferences");
        String pattern = call.getString("pattern");

        try {
            List<String> keys;

            switch (storage) {
                case "secure":
                    if (encryptedStorage == null) {
                        call.reject("Encrypted storage not available");
                        return;
                    }
                    keys = encryptedStorage.keys(pattern);
                    break;
                case "sqlite": {
                    SQLiteStorage store = getSqliteStore(call.getString("database", DEFAULT_DATABASE));
                    if (store == null) {
                        call.reject("SQLite storage not available");
                        return;
                    }
                    keys = store.keys(call.getString("table", DEFAULT_TABLE), pattern);
                    break;
                }
                case "filesystem":
                    if (filesystemStorage == null) {
                        call.reject("Filesystem storage not available");
                        return;
                    }
                    keys = filesystemStorage.keys(pattern);
                    break;
                case "preferences":
                default:
                    if (sharedPrefsStorage == null) {
                        call.reject("Preferences storage not available");
                        return;
                    }
                    keys = sharedPrefsStorage.keys(pattern);
                    break;
            }

            JSObject result = new JSObject();
            result.put("keys", new JSONArray(keys));
            call.resolve(result);
        } catch (Exception e) {
            call.reject("Failed to get keys", e);
        }
    }

    /**
     * Get storage size information.
     *
     * When `detailed` is true the result also carries a byte breakdown
     * { detailed: { keys, values, metadata } }.
     */
    @PluginMethod
    public void size(PluginCall call) {
        String storage = call.getString("storage", "preferences");
        boolean detailed = Boolean.TRUE.equals(call.getBoolean("detailed", false));

        try {
            JSObject result = new JSObject();

            switch (storage) {
                case "secure": {
                    if (encryptedStorage == null) {
                        call.reject("Encrypted storage not available");
                        return;
                    }
                    EncryptedStorage.SizeInfo encryptedSizeInfo = encryptedStorage.size();
                    result.put("total", encryptedSizeInfo.total);
                    result.put("count", encryptedSizeInfo.count);
                    break;
                }
                case "sqlite": {
                    SQLiteStorage store = getSqliteStore(call.getString("database", DEFAULT_DATABASE));
                    if (store == null) {
                        call.reject("SQLite storage not available");
                        return;
                    }
                    SQLiteStorage.SizeInfo sqliteSizeInfo =
                            store.size(call.getString("table", DEFAULT_TABLE), detailed);
                    putSizeInfo(result, sqliteSizeInfo);
                    break;
                }
                case "filesystem": {
                    if (filesystemStorage == null) {
                        call.reject("Filesystem storage not available");
                        return;
                    }
                    SQLiteStorage.SizeInfo fsSizeInfo = filesystemStorage.size(detailed);
                    putSizeInfo(result, fsSizeInfo);
                    break;
                }
                case "preferences":
                default: {
                    if (sharedPrefsStorage == null) {
                        call.reject("Preferences storage not available");
                        return;
                    }
                    SharedPreferencesStorage.SizeInfo prefsSizeInfo = sharedPrefsStorage.size();
                    result.put("total", prefsSizeInfo.total);
                    result.put("count", prefsSizeInfo.count);
                    break;
                }
            }

            call.resolve(result);
        } catch (Exception e) {
            call.reject("Failed to get size", e);
        }
    }

    /**
     * Populate a size result, including the detailed byte breakdown when the
     * {@link SQLiteStorage.SizeInfo} carries one.
     */
    private void putSizeInfo(JSObject result, SQLiteStorage.SizeInfo info) {
        result.put("total", info.total);
        result.put("count", info.count);
        if (info.detailed) {
            JSObject breakdown = new JSObject();
            breakdown.put("keys", info.keysBytes);
            breakdown.put("values", info.valuesBytes);
            breakdown.put("metadata", info.metadataBytes);
            result.put("detailed", breakdown);
        }
    }

    /**
     * Query SQLite-backed storage.
     * Matches the optional `query` method in the JS contract: resolves
     * { results: [{ key, value }] } where `value` is the full StorageValue
     * wrapper for each non-expired row, so the JS SqliteAdapter can apply the
     * condition in one round-trip without re-fetching each key.
     */
    @PluginMethod
    public void query(PluginCall call) {
        String storage = call.getString("storage", "sqlite");
        if (!"sqlite".equals(storage)) {
            call.reject("Query is only supported for the 'sqlite' storage type");
            return;
        }
        SQLiteStorage store = getSqliteStore(call.getString("database", DEFAULT_DATABASE));
        if (store == null) {
            call.reject("SQLite storage not available");
            return;
        }

        try {
            List<Map<String, Object>> rows = store.query(call.getString("table", DEFAULT_TABLE));
            JSArray results = new JSArray();
            for (Map<String, Object> row : rows) {
                JSObject obj = new JSObject();
                obj.put("key", row.get("key"));
                Object value = row.get("value");
                if (value != null) {
                    obj.put("value", value);
                }
                results.put(obj);
            }
            JSObject result = new JSObject();
            result.put("results", results);
            call.resolve(result);
        } catch (Exception e) {
            call.reject("Failed to query SQLite", e);
        }
    }

    /**
     * Delete every expired row for a SQLite store in one statement and resolve
     * { removed: <count> }. The JS adapter calls this on its TTL cleanup tick,
     * since keys()/query() now exclude expired rows in SQL (no per-key delete).
     */
    @PluginMethod
    public void cleanupExpired(PluginCall call) {
        String storage = call.getString("storage", "sqlite");
        if (!"sqlite".equals(storage)) {
            call.reject("cleanupExpired is only supported for the 'sqlite' storage type");
            return;
        }
        SQLiteStorage store = getSqliteStore(call.getString("database", DEFAULT_DATABASE));
        if (store == null) {
            call.reject("SQLite storage not available");
            return;
        }
        try {
            int removed = store.cleanupExpired(call.getString("table", DEFAULT_TABLE));
            JSObject result = new JSObject();
            result.put("removed", removed);
            call.resolve(result);
        } catch (Exception e) {
            call.reject("Failed to clean up expired SQLite rows", e);
        }
    }

    /**
     * Android-specific: read an encrypted preference.
     * Resolves { value: unknown }. Honors an optional `fileName`.
     */
    @PluginMethod
    public void getEncryptedPreference(PluginCall call) {
        String key = call.getString("key");
        if (key == null) {
            call.reject("Key is required");
            return;
        }
        String fileName = call.getString("fileName");

        try {
            EncryptedStorage store = resolveEncryptedStore(fileName);
            if (store == null) {
                call.reject("Encrypted storage not available");
                return;
            }
            JSObject result = new JSObject();
            result.put("value", store.get(key));
            call.resolve(result);
        } catch (Exception e) {
            call.reject("Failed to read encrypted preference", e);
        }
    }

    /**
     * Android-specific: write an encrypted preference. Honors an optional
     * `fileName`.
     */
    @PluginMethod
    public void setEncryptedPreference(PluginCall call) {
        String key = call.getString("key");
        if (key == null) {
            call.reject("Key is required");
            return;
        }
        Object value = call.getData().opt("value");
        String fileName = call.getString("fileName");

        try {
            EncryptedStorage store = resolveEncryptedStore(fileName);
            if (store == null) {
                call.reject("Encrypted storage not available");
                return;
            }
            if (!store.set(key, value)) {
                call.reject("Failed to persist encrypted preference");
                return;
            }
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to write encrypted preference", e);
        }
    }

    /**
     * Returns the default encrypted store, or builds a file-specific one when a
     * custom file name is supplied.
     */
    private EncryptedStorage resolveEncryptedStore(String fileName) throws Exception {
        if (fileName == null || fileName.isEmpty()) {
            return encryptedStorage;
        }
        return new EncryptedStorage(getContext(), fileName);
    }
}
